### PR TITLE
Add node Taint remediation

### PR DESCRIPTION
**Pull request by mshitrit**
This PR is the first attempt to implementing NEC requirments defined in  [this](https://issues.redhat.com/browse/ECOPROJECT-777) ticket.

**Implementation high level details**
 
 Added a new remediation strategy "Add Taint" (the following is relevant for this strategy):
1. Upon CR detection set NoExucte taint on the Node & add finalizer to the CR
2. This should trigger the actual remediation (implemented in the API Server)
3. When remediation is successful NHC will try to delete the CR
4. When CR is attempted to be deleted, remove the finalizer from the CR & the taint from the Node

### CODE REVIEW

File: controllers/selfnoderemediation_controller.go [hunk 1]
```diff
@@ -664,3 +701,37 @@ func (r *SelfNodeRemediationReconciler) updateSnrStatusLastError(snr *v1alpha1.S
...
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *SelfNodeRemediationReconciler) removeOutOfServiceTaint(node *v1.Node) error {
+	patch := client.MergeFrom(node.DeepCopy())
+	var taintIndex int
+	taintFound := false
+	for index, taint := range node.Spec.Taints {
+		if reflect.DeepEqual(taint, *NodeNoExecuteTaint) {
```
Comment (slintes) [inline controllers/selfnoderemediation_controller.go:720]: maybe use k8s.io/api/core/v1/taint.go MatchTaint

**Review by slintes — COMMENTED** [review]
- This feature should be documented somewhere, especially noting when it works and when not (k8s/ocp version, feature status, feature gates...)
- More important: check the note about the taint here: https://kubernetes.io/docs/concepts/architecture/nodes/#non-graceful-node-shutdown
> Before adding the taint node.kubernetes.io/out-of-service , it should be verified that the node is already in shutdown or power off state (not in the middle of restarting).

At the moment shutdown is completely missing in the implementation. And "not in the middle of restarting" does not align with our reboot-and-guess-when-it's-done approach of other strategies. We need a safe way to shut the node down, or to detect that it aleady is. But then we can't switch on, which contradicts our goal of node recovery... 🤔 See also discussion in https://github.com/medik8s/self-node-remediation/issues/17

**Comment by beekhof** [comment]
Closing the loop here, the expiry of the timeout implies shutdown - with a confidence that varies based on the users' environment and configuration (hardware watchdog, software watchdog, software reboot)

File: controllers/selfnoderemediation_controller.go [hunk 2]
```diff
@@ -370,6 +378,35 @@ func (r *SelfNodeRemediationReconciler) remediateWithNodeDeletion(snr *v1alpha1.
...
+func (r *SelfNodeRemediationReconciler) remediateByAddingTaint(snr *v1alpha1.SelfNodeRemediation) (ctrl.Result, error) {
+	node, err := r.getNodeFromSnr(snr)
+	if err != nil {
+		r.logger.Error(err, "failed to get node", "node name", snr.Name)
+		return ctrl.Result{}, err
+	}
+
+	if !controllerutil.ContainsFinalizer(snr, SNRFinalizer) {
+		return r.addFinalizer(snr)
+	}
```
Comment (beekhof) [inline controllers/selfnoderemediation_controller.go:390]: Shouldn't that happen for all strategies before this function gets called?

Comment (beekhof) [reply controllers/selfnoderemediation_controller.go:390]: Even more concerning... 

```
        wasRebooted, timeLeft := r.wasNodeRebooted(snr)
	if !wasRebooted {
		return ctrl.Result{RequeueAfter: timeLeft}, nil
	}

```

**MUST** also be part of ***every*** strategy

**Comment by mshitrit** [comment]
Closing this PR.
Implementation will be done in smaller segments (possibly by different teams), addressing the issues that were brought up.

**Pull request closed without merging**